File: src-tauri/src/analytics/buckets.rs
// Time bucket helpers — granularity derivation and key/label formatting.

use chrono::{Datelike, NaiveDate, Timelike};
use serde::{Deserialize, Serialize};

use super::types::TimeBucketUsage;

/// Bucket granularity derived from the requested day count.
///   1 day          → Hourly  (24 buckets)
///   2–14 days      → Daily
///   15–56 days     → Weekly  (Mon-based weeks)
///   57–90 days     → Monthly
///   >90            → rejected by the frontend
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum BucketGranularity {
    Hourly,
    Daily,
    Weekly,
    Monthly,
}

pub fn granularity_for_days(days: u32) -> BucketGranularity {
    match days {
        0..=2 => BucketGranularity::Hourly,
        3..=14 => BucketGranularity::Daily,
        15..=56 => BucketGranularity::Weekly,
        _ => BucketGranularity::Monthly,
    }
}

pub fn day_key(ts_ms: f64) -> String {
    let secs = (ts_ms / 1000.0) as i64;
    let dt = chrono::DateTime::from_timestamp(secs, 0).unwrap_or_default();
    format!("{:04}-{:02}-{:02}", dt.year(), dt.month(), dt.day())
}

pub fn hour_key(ts_ms: f64) -> String {
    let secs = (ts_ms / 1000.0) as i64;
    let dt = chrono::DateTime::from_timestamp(secs, 0).unwrap_or_default();
    format!(
        "{:04}-{:02}-{:02}-{:02}",
        dt.year(),
        dt.month(),
        dt.day(),
        dt.hour()
    )
}

pub fn day_label(date: &NaiveDate) -> String {
    date.format("%b %-d").to_string()
}

pub fn hour_label(h: u32) -> String {
    match h {
        0 => "12 AM".into(),
        h if h < 12 => format!("{h} AM"),
        12 => "12 PM".into(),
        h => format!("{} PM", h - 12),
    }
}

pub fn week_key(ts_ms: f64) -> String {
    let secs = (ts_ms / 1000.0) as i64;
    let dt = chrono::DateTime::from_timestamp(secs, 0).unwrap_or_default();
    let naive = dt.date_naive();
    let monday = naive - chrono::Duration::days(naive.weekday().num_days_from_monday() as i64);
    format!("{:04}-W{:02}", monday.year(), monday.iso_week().week())
}

pub fn week_label(monday: &NaiveDate) -> String {
    let sunday = *monday + chrono::Duration::days(6);
    if monday.month() == sunday.month() {
        format!("{} {}-{}", monday.format("%b"), monday.day(), sunday.day())
    } else {
        format!("{} - {}", monday.format("%b %-d"), sunday.format("%b %-d"))
    }
}

pub fn month_key(ts_ms: f64) -> String {
    let secs = (ts_ms / 1000.0) as i64;
    let dt = chrono::DateTime::from_timestamp(secs, 0).unwrap_or_default();
    format!("{:04}-{:02}", dt.year(), dt.month())
}

pub fn month_label(year: i32, month: u32) -> String {
    let d = NaiveDate::from_ymd_opt(year, month, 1).unwrap_or_default();
    d.format("%b %Y").to_string()
}

pub fn make_empty_bucket(key: String, label: String) -> TimeBucketUsage {
    TimeBucketUsage {
        key,
        label,
        total_tokens: 0,
        input_tokens: 0,
        output_tokens: 0,
        cache_read_tokens: 0,
        cost_usd: 0.0,
        session_count: 0,
    }
}

pub fn bucket_key_for(granularity: BucketGranularity, ts_ms: f64) -> String {
    match granularity {
        BucketGranularity::Hourly => hour_key(ts_ms),
        BucketGranularity::Daily => day_key(ts_ms),
        BucketGranularity::Weekly => week_key(ts_ms),
        BucketGranularity::Monthly => month_key(ts_ms),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_day_key() {
        let ts_ms = 1705320000.0 * 1000.0;
        assert_eq!(day_key(ts_ms), "2024-01-15");
    }

    #[test]
    fn test_hour_key() {
        let ts_ms = 1705320000.0 * 1000.0;
        assert_eq!(hour_key(ts_ms), "2024-01-15-12");
    }

    #[test]
    fn test_hour_label_midnight() {
        assert_eq!(hour_label(0), "12 AM");
    }

    #[test]
    fn test_hour_label_morning() {
        assert_eq!(hour_label(9), "9 AM");
    }

    #[test]
    fn test_hour_label_noon() {
        assert_eq!(hour_label(12), "12 PM");
    }

    #[test]
    fn test_hour_label_afternoon() {
        assert_eq!(hour_label(15), "3 PM");
    }

    #[test]
    fn test_granularity_hourly() {
        assert!(matches!(granularity_for_days(1), BucketGranularity::Hourly));
        assert!(matches!(granularity_for_days(2), BucketGranularity::Hourly));
    }

    #[test]
    fn test_granularity_daily() {
        assert!(matches!(granularity_for_days(3), BucketGranularity::Daily));
        assert!(matches!(granularity_for_days(14), BucketGranularity::Daily));
    }

    #[test]
    fn test_granularity_weekly() {
        assert!(matches!(granularity_for_days(15), BucketGranularity::Weekly));
        assert!(matches!(granularity_for_days(56), BucketGranularity::Weekly));
    }

    #[test]
    fn test_granularity_monthly() {
        assert!(matches!(granularity_for_days(57), BucketGranularity::Monthly));
        assert!(matches!(granularity_for_days(90), BucketGranularity::Monthly));
    }

    #[test]
    fn test_month_key() {
        let ts_ms = 1705320000.0 * 1000.0;
        assert_eq!(month_key(ts_ms), "2024-01");
    }

    #[test]
    fn test_month_label() {
        assert_eq!(month_label(2024, 1), "Jan 2024");
        assert_eq!(month_label(2024, 12), "Dec 2024");
    }

    #[test]
    fn test_make_empty_bucket() {
        let b = make_empty_bucket("key".to_string(), "label".to_string());
        assert_eq!(b.key, "key");
        assert_eq!(b.total_tokens, 0);
        assert_eq!(b.session_count, 0);
    }
}

File: internal/analytics/buckets.go
// Bucket granularity helpers — mirrors src-tauri/src/analytics/buckets.rs.
package analytics

import (
	"fmt"
	"time"
)

// BucketGranularity mirrors buckets::BucketGranularity.
// JSON values must match the serde camelCase enum variants the frontend expects.
type BucketGranularity string

const (
	GranularityHourly  BucketGranularity = "hourly"
	GranularityDaily   BucketGranularity = "daily"
	GranularityWeekly  BucketGranularity = "weekly"
	GranularityMonthly BucketGranularity = "monthly"
)

// GranularityForDays derives the bucket granularity from the day count.
// Mirrors buckets::granularity_for_days.
func GranularityForDays(days uint32) BucketGranularity {
	switch {
	case days <= 2:
		return GranularityHourly
	case days <= 14:
		return GranularityDaily
	case days <= 56:
		return GranularityWeekly
	default:
		return GranularityMonthly
	}
}

// BucketKeyFor returns the bucket key string for a timestamp (ms) given granularity.
// Mirrors buckets::bucket_key_for.
func BucketKeyFor(g BucketGranularity, tsMs float64) string {
	t := msToTime(tsMs)
	switch g {
	case GranularityHourly:
		return hourKey(t)
	case GranularityDaily:
		return dayKey(t)
	case GranularityWeekly:
		return weekKey(t)
	default:
		return monthKey(t)
	}
}

func msToTime(ms float64) time.Time {
	secs := int64(ms / 1000.0)
	nsec := int64((ms/1000.0-float64(secs))*1e9)
	return time.Unix(secs, nsec).UTC()
}

func dayKey(t time.Time) string {
	return fmt.Sprintf("%04d-%02d-%02d", t.Year(), t.Month(), t.Day())
}

func hourKey(t time.Time) string {
	return fmt.Sprintf("%04d-%02d-%02d-%02d", t.Year(), t.Month(), t.Day(), t.Hour())
}

func weekKey(t time.Time) string {
	// ISO week Mon-based, matching Rust chrono.
	monday := isoMonday(t)
	y, w := monday.ISOWeek()
	return fmt.Sprintf("%04d-W%02d", y, w)
}

func monthKey(t time.Time) string {
	return fmt.Sprintf("%04d-%02d", t.Year(), t.Month())
}

// DayLabel formats a date as "Jan 2" (%-d in Rust). Mirrors buckets::day_label.
func DayLabel(t time.Time) string {
	return fmt.Sprintf("%s %d", t.Format("Jan"), t.Day())
}

// HourLabel formats a 0-23 hour as "12 AM", "9 AM", "12 PM", "3 PM".
// Mirrors buckets::hour_label.
func HourLabel(h int) string {
	switch {
	case h == 0:
		return "12 AM"
	case h < 12:
		return fmt.Sprintf("%d AM", h)
	case h == 12:
		return "12 PM"
	default:
		return fmt.Sprintf("%d PM", h-12)
	}
}

// WeekLabel formats a Monday date as "Jan 2-8" or "Jan 31 - Feb 6".
// Mirrors buckets::week_label.
func WeekLabel(monday time.Time) string {
	sunday := monday.AddDate(0, 0, 6)
	if monday.Month() == sunday.Month() {
		return fmt.Sprintf("%s %d-%d", monday.Format("Jan"), monday.Day(), sunday.Day())
	}
	return fmt.Sprintf("%s %d - %s %d",
		monday.Format("Jan"), monday.Day(),
		sunday.Format("Jan"), sunday.Day())
}

// MonthLabel formats year+month as "Jan 2024". Mirrors buckets::month_label.
func MonthLabel(year int, month time.Month) string {
	return fmt.Sprintf("%s %d", month.String()[:3], year)
}

// MakeEmptyBucket returns a zero-valued TimeBucketUsage. Mirrors buckets::make_empty_bucket.
func MakeEmptyBucket(key, label string) TimeBucketUsage {
	return TimeBucketUsage{
		Key:             key,
		Label:           label,
		TotalTokens:     0,
		InputTokens:     0,
		OutputTokens:    0,
		CacheReadTokens: 0,
		CostUSD:         0.0,
		SessionCount:    0,
	}
}

// isoMonday returns the Monday of the ISO week containing t.
func isoMonday(t time.Time) time.Time {
	// time.Weekday: 0=Sun, 1=Mon..6=Sat; ISO Mon=0..Sun=6
	wd := int(t.Weekday())
	if wd == 0 {
		wd = 7
	}
	return t.AddDate(0, 0, -(wd - 1)).Truncate(24 * time.Hour)
}
